### ISSUE TITLE
Allow to configure IO awaited Tasks to complete on the IOCP worker thread.

### ISSUE THREAD

**Issue by clrjunkie** [dotnet/corefx#4834]
For background info see @stephentoub excellent post: 

http://blogs.msdn.com/b/pfxteam/archive/2012/01/20/10259049.aspx

Key relevant quote:

"If the context comes back as null, then the continuation could run “anywhere”. I put anywhere in quotes because obviously the continuation can’t run “anywhere,” but logically you can think of it like that… it’ll either end up running on the same thread that completed the awaited task, or it’ll end up running in the ThreadPool."

What I noticed in .NET FW 4.6 is that HttpListener completions _sometimes_ get posted to the ThreadPool _Work Queue_ instead of being directly completed on the IOCP worker thread itself.

The way I manage to avoid this unnecessarily context switch is by "overriding" the Task code policy on how to dispatch the completion:

var currentContext = SynchronizationContext.Current;
try
{
    SynchronizationContext.SetSynchronizationContext(new NoOpSynchronizationContext());

```
await listener.GetContextAsync()
```

}
finally
{
    SynchronizationContext.SetSynchronizationContext(currentContext);
}

public class NoOpSynchronizationContext : SynchronizationContext
{
    public override void Post(SendOrPostCallback d, object state)
    {
        d(state);
    }
}

Needless to say this is boilerplate code one would want to avoid when awaiting tasks of libs using IOCP.

**Comment by stephentoub** [comment]
> What I noticed in .NET FW 4.6 is that HttpListener completions sometimes get posted to the ThreadPool Work Queue instead of being directly completed on the IOCP worker thread itself.

Your example isn't using ConfigureAwait(false).  Is that just a typo?  With the default (which is ConfigureAwait(true), if there is a SynchronizationContext, you'll post back to it, so if for example there was a default SynchronizationContext set on the current thread, then you'd post back to the worker pool.

>  completions sometimes get posted to the ThreadPool Work Queue

How often?

**Comment by clrjunky** [comment]
I intentionally didn’t use ConfigureAwait(false) in order to have the task code post to my custom synchronization context so the callback will simply pass through and continue executing on the completion thread.

I did this after checking if ConfigureAwait(false) would prevent the callback from executing on the thread pool, which it didn’t.

I saw this behavior in a console application and I attempted to debug it using the reference source symbols but I couldn’t step through all the code... 

Unfortunately I didn’t find this to be deterministic ..  I would say it normally completes on the same thread for the first or second request and than the callback starts to get posted to the thread pool.

One correction, I observed this on .net 4.5.2 (win7/vs2013)

**Comment by stephentoub** [comment]
> I intentionally didn’t use ConfigureAwait(false) in order to have the task code post to my custom synchronization context so the callback will simply pass through and continue executing on the completion thread. I did this after checking if ConfigureAwait(false) would prevent the callback from executing on the thread pool, which it didn’t.

This is confusing to me.  The only way that should behave as you suggest is if there's already a non-default SynchronizationContext set on the thread.  So, when this repros, what's the type of SynchronizationContext.Current on the thread calling GetContextAsync (before you change it), and what's the type of SynchronizationContext.Current inside of the call to your Post method?

I'm also a bit confused by the title of this issue.  Are you asking for a built-in way to say that any work items that would normally be queued to the worker pool should instead be queued to the IOCP pool, or are you filing a bug on HttpListener that some of the tasks returned from GetContextAsync complete on a worker thread rather than on an IOCP thread? Or something else?

Thanks.

**Comment by kouvel** [comment]
If you could provide a repro and what you expect to see, that would be greatly useful as well. Thanks!

**Comment by clrjunkie** [comment]
@stephentoub, @kouvel 
 >So, when this repros, what's the type of SynchronizationContext.Current on the thread calling GetContextAsync (before you change it), and what's the type of SynchronizationContext.Current inside of the call to your Post method?

Since I opened this issue about a month ago my environment has drastically changed, currently on Win10/VS2015.  I noticed the issue on Win7/VS2013/.NET 4.52 in Debug mode. I can't revert to that config but I'll report if I'll able to repro it again in the current environment.
  
 @stephentoub
>The only way that should behave as you suggest is if there's already a non-default SynchronizationContext set on the thread. 

This is key. I'm not familiar with the compiler implementation that generates the underlying Async support code. It was my understanding from your article that it is "undefined" to the user which thread will actually execute the continuation by default. Furthermore, the term “ThreadPool” is rather ambiguous in .NET because it can be the Managed implementation or the Native one that’s created by BindHandle >BindIOCompletionCallbackNative

"If the context comes back as null, then the continuation could run “anywhere”. I put anywhere in quotes because obviously the continuation can’t run “anywhere,” but logically you can think of it like that… it’ll either end up running on the same thread that completed the awaited task, or it’ll end up running in the ThreadPool."

If indeed the presence or lack of presence of a SynchronizationContext on the thread that started the Async operation serves as the *only* selection criteria as to which thread will execute the continuation, then I agree this is VERY strange.

Aside from checking the state of the SynchronizationContext as you suggested, can you please point me to the exact class and line number (in core Clr/.net 4.6) where this decision is taking place so I can put a breakpoint?
 
>Are you asking for a built-in way to say that any work items that would normally be queued to the worker pool should instead be queued to the IOCP pool, or are you filing a bug on HttpListener that some of the tasks returned from GetContextAsync complete on a worker thread rather than on an IOCP thread? Or something else?

The reasons I opened this issue with reference to HttpListener are:

1. I noticed this when using HttpListener.
2. I'm pretty familiar with HttpListener internal implementation and I haven't noticed anything special that causes it to change the SyncContext.

Simply put, the whole point of this issue is to alert that when an IO Callback executes there is no need to resubmit the continuation to the .NET Managed ThreadPool, which will only incur an unnecessary context switch, but rather keep executing the continuation on the IO Thread.
Whether my experience with HttpListener actually represents such non-deterministic deficiency is yet to be verified, but did occur in my previous environment.

**Comment by clrjunkie** [comment]
@stephentoub can you please clarify the behavior.

**Comment by kouvel** [comment]
> ... the presence or lack of presence of a SynchronizationContext on the thread that started the Async operation serves as the only selection criteria as to which thread will execute the continuation ...

Yes, plus `TaskContinuationOptions.ExecuteSynchronously` and `TaskCreationOptions.RunContinuationsAsynchronously` can affect where a continuation runs. See calls to `ThreadPool.UnsafeQueueCustomWorkItem` and the conditions under which that happens.

It's not very clear from the information we have so far that there is an actual issue here. Closing, please reopen if you have new info that is likely to make this actionable.

**Comment by davidfowl** [comment]
@clrjunkie are you trying to make `GetContextAsync` run continuations on the IO thread?

**Comment by clrjunkie** [comment]
@davidfowl Yes. and as stated at the beginning of this issue it's tedious and code bloat with SynchronizationContext most of the time the next operation is also an enqueue to IOCP (e.g Database socket call) so there is no need to switch to another thread for that.. quite frankly I don't see any obvious use-case that would justify a thread switch .. Windows thread pool will release another thread if an IO thread blocks and if I recall correctly it maintains 1000 of those just waiting to kick-in...

@kouvel This issue is open more than year... What exact information is missing?

**Comment by stephentoub** [comment]
> can you please clarify the behavior.

Somehow I missed this question when it was added last year. Apologies.

You can see the implementation here.  This code determines what kind of continuation to create based on the value of ConfigureAwait's bool and the current SynchronizationContext and TaskScheduler:
https://source.dot.net/#System.Private.CoreLib/src/System/Threading/Tasks/Task.cs,2579
and if there was a captured SynchronizationContext, this code determines whether to run the continuation synchronously as part of the antecedent's completion or asynchronously:
https://source.dot.net/#System.Private.CoreLib/src/System/Threading/Tasks/TaskContinuation.cs,401
That canInlineContinuationTask bool is set here:
https://source.dot.net/#System.Private.CoreLib/src/System/Threading/Tasks/Task.cs,3230

**Comment by davidfowl** [comment]
@clrjunkie Unfortunately there's no way to guarantee that behavior unless the underlying system supports it natively and gives you ways to specify where callbacks run.
